### PR TITLE
fish-lsp: init at 1.0.8-1

### PR DESCRIPTION
**Pull request by petertriho**
## Description of changes

Adds [`fish-lsp`](https://github.com/ndonfris/fish-lsp) https://github.com/NixOS/nixpkgs/issues/305493

Uses `yarnConfigHook` and `yarnBuildHook` instead of `mkYarnPackage`

Supersedes https://github.com/NixOS/nixpkgs/pull/320463 https://github.com/NixOS/nixpkgs/pull/330320

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by petertriho** [comment]
Note sure why both `passthru.tests` are stuck. Would appreciate help with this

**Comment by petertriho** [comment]
Result of `nixpkgs-review pr 330996` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>fish-lsp</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

**Comment by petertriho** [comment]
Looks like all tests have passed. Is there anything else I should do to get this merged?

**Comment by tye-exe** [comment]
Thanks for putting in the work to do this, i've just spent the last few hours trying to do it myself before i found this. Does anyone know the timeline until this will be merged?

**Comment by Scrumplex** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 330996`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>fish-lsp</li>
  </ul>
</details>

File: pkgs/by-name/fi/fish-lsp/package.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    runHook preInstall
+
+    mkdir -p $out/share/fish-lsp
+    cp -r . $out/share/fish-lsp
+
+    makeWrapper ${lib.getExe nodejs} "$out/bin/fish-lsp" \
+      --add-flags "$out/share/fish-lsp/out/cli.js"
+
+    installShellCompletion --cmd fish-lsp \
+      --fish <($out/bin/fish-lsp complete --fish)
```
Comment (Scrumplex) [inline pkgs/by-name/fi/fish-lsp/package.nix:60]: When cross compiling, we (the build platform) *might* not be able to run the final program (built for the host platform). This should be a conditional like this:

```suggestion
    ${lib.optionalString stdenv.buildPlatform.canExecute stdenv.hostPlatform ''
      installShellCompletion --cmd fish-lsp \
        --fish <($out/bin/fish-lsp complete --fish)
    ''}
```

File: pkgs/by-name/fi/fish-lsp/package.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  passthru.updateScript = nix-update-script { };
+
+  meta = with lib; {
+    description = "LSP implementation for the fish shell language";
+    homepage = "https://github.com/ndonfris/fish-lsp";
+    license = licenses.mit;
+    mainProgram = "fish-lsp";
+    maintainers = with maintainers; [ petertriho ];
+    platforms = platforms.unix;
+  };
```
Comment (Scrumplex) [inline pkgs/by-name/fi/fish-lsp/package.nix:76]: nit:
```suggestion
  meta = {
    description = "LSP implementation for the fish shell language";
    homepage = "https://github.com/ndonfris/fish-lsp";
    license = lib.licenses.mit;
    mainProgram = "fish-lsp";
    maintainers = with lib.maintainers; [ petertriho ];
    platforms = lib.platforms.unix;
  };
```

File: pkgs/by-name/fi/fish-lsp/package.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  nativeBuildInputs = [
+    yarnBuildHook
+    yarnConfigHook
+    npmHooks.npmInstallHook
+    nodejs
+    installShellFiles
+    makeWrapper
+    # fish-lsp dependencies
+    fish
+    which
```
Comment (Scrumplex) [inline pkgs/by-name/fi/fish-lsp/package.nix:41]: are these needed at build time or at runtime?

Comment (petertriho) [reply pkgs/by-name/fi/fish-lsp/package.nix:41]: Hi @Scrumplex,

Thank you for having look at the PR.

- `fish` is required for the `sh:build-wasm` script which runs https://github.com/ndonfris/fish-lsp/blob/master/scripts/build-fish-wasm.fish 
Error without `fish`
```
> Running phase: buildPhase
       > Executing yarnBuildHook
       > yarn run v1.22.22
       > $ run-s sh:build-time sh:build-wasm
       > warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
       > $ fish ./scripts/build-time.fish
       > /bin/sh: fish: not found
       > error Command failed with exit code 127.
       > info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       > ERROR: "sh:build-time" exited with 127.
       > error Command failed with exit code 1.
```
- `which` is required to generate the completion. See https://github.com/ndonfris/fish-lsp/blob/master/src/utils/builtins.ts (`findShell` spawns `which fish`)
Error without `which`
```
       > Running phase: installPhase
       > /bin/sh: which: not found
       > /nix/store/aam03bwd19cadd1ad6qw26gr7h7hkh7m-fish-lsp-unstable-2024-07-26/share/fish-lsp/out/utils/builtins.js:88
       >     return result.stdout.toString().split('\n');
       >                          ^
       >
       > TypeError: Cannot read properties of null (reading 'toString')
       >     at createFunctionNamesList (/nix/store/aam03bwd19cadd1ad6qw26gr7h7hkh7m-fish-lsp-unstable-2024-07-26/share/fish-lsp/out/utils/builtins.js:88:26)
       >     at Object.<anonymous> (/nix/store/aam03bwd19cadd1ad6qw26gr7h7hkh7m-fish-lsp-unstable-2024-07-26/share/fish-lsp/out/utils/builtins.js:90:29)
       >     at Module._compile (node:internal/modules/cjs/loader:1469:14)
       >     at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
       >     at Module.load (node:internal/modules/cjs/loader:1288:32)
       >     at Module._load (node:internal/modules/cjs/loader:1104:12)
       >     at Module.require (node:internal/modules/cjs/loader:1311:19)
       >     at require (node:internal/modules/helpers:179:18)
       >     at Object.<anonymous> (/nix/store/aam03bwd19cadd1ad6qw26gr7h7hkh7m-fish-lsp-unstable-2024-07-26/share/fish-lsp/out/utils/translation.js:38:20)
       >     at Module._compile (node:internal/modules/cjs/loader:1469:14)
       >
       > Node.js v20.17.0
       > installShellCompletion: installed shell completion file `/nix/store/aam03bwd19cadd1ad6qw26gr7h7hkh7m-fish-lsp-unstable-2024-07-26/share/fish/vendor_completions.d/fish-lsp.fish' does not exist or has zero size
```

File: pkgs/by-name/fi/fish-lsp/package.nix (hunk 1 above)
Comment (petertriho) [reply pkgs/by-name/fi/fish-lsp/package.nix]: Fixed

File: pkgs/by-name/fi/fish-lsp/package.nix (hunk 2 above)
Comment (petertriho) [reply pkgs/by-name/fi/fish-lsp/package.nix]: Fixed

File: pkgs/by-name/fi/fish-lsp/package.nix [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/share/fish-lsp
+    cp -r . $out/share/fish-lsp
+
+    makeWrapper ${lib.getExe nodejs} "$out/bin/fish-lsp" \
+      --add-flags "$out/share/fish-lsp/out/cli.js"
+
+    ${lib.optionalString stdenv.buildPlatform.canExecute stdenv.hostPlatform ''
```
Comment (Scrumplex) [inline pkgs/by-name/fi/fish-lsp/package.nix:59]: Forgot parenthesis. Sorry!

```suggestion
    ${lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
```

File: pkgs/by-name/fi/fish-lsp/package.nix (hunk 3 above)
Comment (Scrumplex) [reply pkgs/by-name/fi/fish-lsp/package.nix:41]: I think we should add those packages to the wrapper if they are sometimes used at runtime.

Like this patch:

```diff
diff --git a/pkgs/by-name/fi/fish-lsp/package.nix b/pkgs/by-name/fi/fish-lsp/package.nix
index bc9ceb51ecb8..cefbdbaf366a 100644
--- a/pkgs/by-name/fi/fish-lsp/package.nix
+++ b/pkgs/by-name/fi/fish-lsp/package.nix
@@ -36,9 +36,7 @@ stdenv.mkDerivation rec {
     nodejs
     installShellFiles
     makeWrapper
-    # fish-lsp dependencies
     fish
-    which
   ];
 
   yarnBuildScript = "setup";
@@ -54,7 +52,8 @@ stdenv.mkDerivation rec {
     cp -r . $out/share/fish-lsp
 
     makeWrapper ${lib.getExe nodejs} "$out/bin/fish-lsp" \
-      --add-flags "$out/share/fish-lsp/out/cli.js"
+      --add-flags "$out/share/fish-lsp/out/cli.js" \
+      --prefix PATH : "${lib.makeBinPath [fish which]}"
 
     ${lib.optionalString stdenv.buildPlatform.canExecute stdenv.hostPlatform ''
       installShellCompletion --cmd fish-lsp \
```

Comment (petertriho) [reply pkgs/by-name/fi/fish-lsp/package.nix]: Good idea, fixed

File: pkgs/by-name/fi/fish-lsp/package.nix (hunk 4 above)
Comment (petertriho) [reply pkgs/by-name/fi/fish-lsp/package.nix]: My bad, I should have checked this

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 330996`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>fish-lsp</li>
  </ul>
</details>

**Comment by petertriho** [comment]
@Scrumplex looks like all the tests finally pass. What are the next steps to get this merged?

**Comment by Scrumplex** [comment]
> oops looks like there's 1 more left

Those darwin tests tend to get stuck, sadly. But looking at their output manually shows that it built fine :D

**Pull request merged**